### DISCUSSION TITLE
Are there some smart way to not write so many `include` when having too many defined relations in entities(like nestjs) ?

### DISCUSSION THREAD

**Discussion by toknT** [prisma/prisma#25042] [Q&A] [answered]
### Question

I am using nestjs+graphql.

There are 6 tables in my db and relations liks this:

1. `QuestionMistakeTag`  -(n:m)-> `QuestionMistake`  -(n:1)-> `Question` -(n:1)-> `QuestionBank`-(1?:1)-> `Resource`  -(n:1)-> `Category`
2. `Resource`  -(n:1)-> `User`
3. `QuestionMistake`  -(n:1)-> `User`

After defined the relations in entities.  If access `QuestionMistake`  it need 4 nested `include`.

```ts
findMany(page: number, perPage: number): Promise<QuestionMistake[]> {
    return this.prisma.questionMistake.findMany({
      include: {
        user: true,
        tags: {
          include: {
            tag: true,
          },
        },
        question: {
          include: {
            questionBank: {
              include: {
                resource: { include: { category: true, creator: true } },
              },
            },
          },
        },
      },
      skip: perPage * page,
      take: perPage,
    });
  }
```

Are there some smart way to not write so many `include` ?

### Information about Prisma Schema, Client Queries and Environment (optional)

```sh
$ npx prisma -v
Environment variables loaded from .env
prisma                  : 5.17.0
@prisma/client          : 5.17.0
Computed binaryTarget   : windows
Operating System        : win32
Architecture            : x64
Node.js                 : v20.10.0
Query Engine (Node-API) : libquery-engine 393aa359c9ad4a4bb28630fb5613f9c281cde053 (at node_modules\@prisma\engines\query_engine-windows.dll.node)
Schema Engine           : schema-engine-cli 393aa359c9ad4a4bb28630fb5613f9c281cde053 (at node_modules\@prisma\engines\schema-engine-windows.exe)
Schema Wasm             : @prisma/prisma-schema-wasm 5.17.0-31.393aa359c9ad4a4bb28630fb5613f9c281cde053
Default Engines Hash    : 393aa359c9ad4a4bb28630fb5613f9c281cde053
Studio                  : 0.502.0
```

**Comment by ludralph** [comment, marked as answer]
Hi @toknT 👋 

There isn't a direct way to avoid writing multiple nested `include` statements in Prisma Client when you need to fetch deeply nested relations. 

However, there are some strategies you can consider to optimize and manage your queries better:

1. Using JOINs for Relations: Prisma Client has support for JOINs to resolve relations, which can help optimize the performance of your queries. This feature is available under the[ relationJoins](https://www.prisma.io/docs/orm/prisma-client/queries/relation-queries#relation-load-strategies-preview) preview feature. Enabling this feature can help reduce the number of round-trips to the database, making your queries more efficient.

To enable this feature, you need to add it to your schema.prisma file:
```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["relationJoins"]
}
```
2. Modularizing Include Statements: While this doesn't reduce the number of include statements, you can modularize your include objects to make your code cleaner and more maintainable. For example:
```js
const questionBankInclude = {
  resource: { include: { category: true, creator: true } },
};

const questionInclude = {
  questionBank: { include: questionBankInclude },
};

const questionMistakeInclude = {
  user: true,
  tags: { include: { tag: true } },
  question: { include: questionInclude },
};

findMany(page: number, perPage: number): Promise<QuestionMistake[]> {
  return this.prisma.questionMistake.findMany({
    include: questionMistakeInclude,
    skip: perPage * page,
    take: perPage,
  });
}
```

3. Optimizing Queries: Ensure that you are only including the necessary fields and relations to avoid fetching unnecessary data. This can help improve the performance of your queries.

**Reply by toknT** [reply]
Thanks , I will try to define `Include` in entities
